### DISCUSSION TITLE
Feeding spikes per timestep and learning delays

### DISCUSSION THREAD

**Discussion by Huizerd** [lava-nc/lava-dl#19] [Q&A] [answered]
As discussed during the Lava developer forum:

I have two questions regarding the feeding of spikes and the learning of delays. Looking at the tutorial notebooks and the [code](https://github.com/lava-nc/lava-dl/blob/main/src/lava/lib/dl/slayer/axon/delay.py), it seems everything is built around feeding sequences of spikes instead of spikes per timestep, and doing convolutions over these sequences to get current, voltage, etc.

1. I can presumably send spikes per timestep by using a sequence length of 1, but will this limit any of SLAYER's functionalities?
2. More specifically, for the learning of delays, the code shows that spikes and gradients are shifted across time. Does this mean that I can only learn delays up to the length of the sequence because there's no gradient propagation outside, hence disabling delay learning for sending spikes per timestep?

Reply by @bamsumit:

Yes, when we send spikes/inputs one timestep at a time, then the delay functionality of slayer are limited. You are right, at the moment we can learn an effective delay up to the length of sequence. The efficiency of SLAYER is through processing spike train in bulk. This gives execution speed advantage with custom cuda accelerated modules.

It is in principle possible to extend delay learning beyond the sequence length by maintaining forward and backward buffer.

Follow-up question:

Do you have any pointers on how one would go about this, and integrate it nicely with autograd?

**Comment by bamsumit** [comment, marked as answer]
Before we go into that direction, what is preventing you to collect the spikes for a some amount of time and process that during training? The spike train is needed for training only. You can always feed spikes one timestep at a time during inference in a lava process or in Loihi hardware.

**Reply by Huizerd** [reply]
Alright that makes sense, yes
